operator/pkg/resources/statefulset_scale.go

```diff
@@ -71,85 +74,134 @@ const (
 
-	targetOrdinal := r.pandaCluster.Status.CurrentReplicas - 1 // Always decommission last node
+	targetOrdinal := npStatus.CurrentReplicas - 1 // Always decommission last node
 	targetBroker, err := r.getBrokerIDForPod(ctx, targetOrdinal)
 	if err != nil {
 		return fmt.Errorf("error getting broker ID for pod with ordinal %d when downscaling cluster: %w", targetOrdinal, err)
 	}
-	nonExistantBroker := int32(-1)
+
 	if targetBroker == nil {
-		// The target pod isn't in the broker list. Just select a non-existing broker for decommission so the next
-		// reconcile loop will succeed.
-		targetBroker = &nonExistantBroker
-	}
-	log.WithValues("ordinal", targetOrdinal, "node_id", targetBroker).Info("start decommission broker")
-	r.pandaCluster.SetDecommissionBrokerID(targetBroker)
-	err = r.Status().Update(ctx, r.pandaCluster)
-	if err != nil {
-		return err
-	}
-	if *targetBroker == nonExistantBroker {
+		// The target pod isn't in the broker list.
+		// Broker ID could be missing for a variety of reasons? If we just ignore
+		// this, and remove the pod anyway and skip decom'ing..we may remove pod w/
+		// ghost broker left. Instead, we only decommission, if both pod and brokerID are present.
+		// This effectively means, our "state machine" requires upscaling to have finished
+		// (pod started + broker registered), before we consider downscaling.
 		return &RequeueAfterError{
 			RequeueAfter: RequeueDuration,
 			Msg:          fmt.Sprintf("the broker for pod with ordinal %d is not registered with the cluster. Requeuing.", targetOrdinal),
 		}
 	}
+
+	podName := fmt.Sprintf("%s-%d", r.LastObservedState.Name, targetOrdinal)
+	log.WithValues("pod", podName, "ordinal", targetOrdinal, "node_id", targetBroker).Info("start decommission broker")
+
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		cluster := &vectorizedv1alpha1.Cluster{}
+		err := r.Get(ctx, types.NamespacedName{
+			Name:      r.pandaCluster.Name,
+			Namespace: r.pandaCluster.Namespace,
+		}, cluster)
+		if err != nil {
+			return err
+		}
+		cluster.SetDecommissionBrokerID(targetBroker)
+		err = r.Status().Update(ctx, cluster)
+		if err == nil {
+			// sync original cluster variable to avoid conflicts on subsequent operations
+			r.pandaCluster.SetDecommissionBrokerID(targetBroker)
+		}
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
+// getDecommissioningPod finds the pod that belongs to *THIS* StatefulSet. If none is found - nil,nil is returned.
+// This may be the case, if the pod belongs to a different STS.
+func (r *StatefulSetResource) getDecommissioningPod(ctx context.Context, brokerID *int32) (*corev1.Pod, error) {
+	var brokerPod *corev1.Pod
+	// Get all pods of *THIS* nodePool
+	podList, err := r.getPodList(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Filter out those where NodeID != DecomNodeID.
+	podList.Items = slices.DeleteFunc(podList.Items, func(pod corev1.Pod) bool {
```

Comment:
Is there any reason for this to not be a range loop with an early return?

```go
for _, pod := range pods {
    if cond {
        return pod, nil
    }
}

return nil, nil // Pod not found
```

Reply:
See https://github.com/redpanda-data/redpanda-operator/pull/188/commits/9deaad0a31b64af7f2280ebbb171f32e8dc66143

Refactored it as suggested, good call. also changed param to non-pointer, which makes much more sense imho.